fairlearn/metrics/_multi_curve_plot.py

```diff
@@ -0,0 +1,328 @@
+
+    x_axis_metric : Callable
+        The (aggregating) metric function for the x-axis
+        The passed metric function must take `y_true, y_pred`, and optionally `sensitive_features`.
+        If the metric is grouped, it must aggregate results. For instance, use
+        `make_derived_metric(metric=balanced_accuracy_score, transform='group_min')`
+        to aggregate the `balanced_accuracy_score`.
+
+    y_axis_metric : Callable
+        The (aggregating) metric function for the y-axis, similar to x_axis_metric.
+        The passed metric function must take `y_true, y_pred`, and optionally `sensitive_features`.
+        If the metric is grouped, it must aggregate results.
+
+    ax : matplotlib.axes.Axes, optional
+        If supplied, the scatter plot is drawn on this Axes object.
+        Else, a new figure with Axes is created.
+
+    axis_labels : bool, tuple
+        If true, add the names of x and y axis metrics. You can also pass a
+        two-tuple of strings to use as axis labels instead.
+
+    point_labels : bool, list
+        If true, annotate text with inferred point labels. These labels are
+        the keys of y_preds if y_preds is a dictionary, else simply the integers
+        0...number of points - 1. You can specify point_labels as a list of
+        labels as well.
+
+    point_labels_position : tuple
+        a two-tuple of numbers that define the offset of the point labels in the
+        x and y direction respectively. The offset value is in data coordinates,
+        not pixels.
+
+    legend : bool
+        If True, add a legend. Legend entries are created by passing the
+        key word argument :code:`label` to calls to this function.
+
+    legend_kwargs : dict
+        Keyword arguments passed to :py:func:`Axes.legend`. For instance,
+        :code:`legend_kwargs={'bbox_to_anchor': (1.03,1), 'loc': 'upper left'}`
+        will create a legend to the right of the Axes (subfigure), or
+        :code:`legend_kwargs={'ncol': 2}` will create two columns in the legend.
+
+    show_plot : bool
+        If true, call pyplot.plot.
+
+    Returns
+    -------
+    ax : matplotlib.axes.Axes
+        The Axes object that was drawn on.
+
+    Notes
+    -----
+    To offer flexibility in stylistic features besides the aforementioned
+    API options, one has at least three options: 1) supply matplotlib arguments
+    to :code:`plot_model_comparison` as you normally
+    would to :code:`matplotlib.axes.Axes.scatter`
+    2) change the style of the returned Axes
+    3) supply an Axes with your own style already applied
+
+    In case no Axes object is supplied, axis labels are
+    automatically inferred from their class name.
+    """  # noqa: E501
+    # --- CHECK DEPENDENCY ---
+    try:
+        import matplotlib.pyplot as plt
+    except ImportError:
+        raise RuntimeError(_MATPLOTLIB_IMPORT_ERROR_MESSAGE)
+
+    # --- SET OR LOAD CONTEXT STATE ---
+    global context_state
```

Comment:
Mhmm... I see what you're doing. I am a bit worried that this is too opaque and could lead to some strange behavior. It might be better if we instead change the return type from matplotlib.axes.Axes to something like [sklearn.utils.Bunch](https://scikit-learn.org/stable/modules/generated/sklearn.utils.Bunch.html) with fields
* `y_true`, `sensitive_features`, `x_axis_metric`, `y_axis_metric`, `ax`.

We could call this returned tuple `context`, and it could be provided as an input argument, as an alternative to `y_true`, `sensitive_features`, `x_axis_metric`, `y_axis_metric`, `ax` (or it could also be dumped directly via `*context`).

It would mean that we depart from the other plot functions (which return an Axes object), but I'd rather have that than a global variable. I think that this is somewhere in between your current approach and making `ax` stateful.

@romanlutz : for context, we would like to address the issue how to avoid the need to keep supplying the same arguments when adding algorithms to a comparison plot (by repeated calls to `plot_model_comparison`).

Reply:
Yeah that's well put @romanlutz! @MiroDudik the bunch seems like a fine suggestion, but how do you feel about a wrapper instead? I'm not sure why, but I like a wrapper a little bit better.

Also, I'll be on vacation next week, so I won't be able to respond

Reply:
FWIW I had thought of the Bunch as a wrapper, but you're right in that we could have a well-defined class for this. The wrapper object is then an instance of that class. The only benefit I see there is that you can have methods on it, but I don't know how heavyweight we want to go with that. I suspect @MiroDudik suggested the Bunch to remain fairly lightweight.

Reply:
@MiroDudik Any comment on this? :)

Reply:
I'm worried about the complexity that is introduced by the stateful object, so I am in favor of going with the simplest route, and simply require the user to always specify the "repeated" arguments in those rare cases when the user is building the comparison plot "one model at a time". I actually don't think that this is going to be a very common use case, that's why I'm suggesting this simpler API (at the cost of occasional requirement to be more verbose).

So--I don't think that we should do Bunch, I don't think we should do a wrapper, and I don't think we should do a global variable. @SeanMcCarren?

Reply:
Sure, I can understand that preference, so let's do that then